File: swift/Sources/DSA/DataStructures/Queue.swift
/// FIFO (First In, First Out) data structure using a linked list.
/// Time: O(1) for all operations
/// Space: O(n)
public struct Queue<T> {
    private class Node {
        var value: T
        var next: Node?

        init(_ value: T) {
            self.value = value
        }
    }

    private var head: Node?
    private var tail: Node?
    private var size = 0

    public init() {}

    public mutating func enqueue(_ value: T) {
        let node = Node(value)
        if let tail = tail {
            tail.next = node
        } else {
            head = node
        }
        tail = node
        size += 1
    }

    public mutating func dequeue() -> T? {
        guard let node = head else { return nil }
        head = node.next
        if head == nil {
            tail = nil
        }
        size -= 1
        return node.value
    }

    public func peek() -> T? {
        head?.value
    }

    public var isEmpty: Bool {
        size == 0
    }

    public var count: Int {
        size
    }

    public mutating func clear() {
        head = nil
        tail = nil
        size = 0
    }
}

File: typescript/src/data-structures/queue.ts
/**
 * Queue implementation using a linked list.
 * Time: O(1) for all operations
 * Space: O(n)
 */

interface QueueNode<T> {
  value: T;
  next: QueueNode<T> | null;
}

export class Queue<T> {
  private head: QueueNode<T> | null = null;
  private tail: QueueNode<T> | null = null;
  private _size: number = 0;

  enqueue(value: T): void {
    const node: QueueNode<T> = { value, next: null };
    if (this.tail) {
      this.tail.next = node;
    } else {
      this.head = node;
    }
    this.tail = node;
    this._size++;
  }

  dequeue(): T | undefined {
    if (!this.head) {
      return undefined;
    }
    const value = this.head.value;
    this.head = this.head.next;
    if (!this.head) {
      this.tail = null;
    }
    this._size--;
    return value;
  }

  peek(): T | undefined {
    return this.head?.value;
  }

  isEmpty(): boolean {
    return this._size === 0;
  }

  size(): number {
    return this._size;
  }

  clear(): void {
    this.head = null;
    this.tail = null;
    this._size = 0;
  }
}
